bmc/connection.go

```diff
@@ -28,67 +29,105 @@ type connectionProviders struct {
-	var metadataLocal Metadata
+func OpenConnectionFromInterfaces(ctx context.Context, timeout time.Duration, providers []interface{}) (opened []interface{}, _ Metadata, err error) {
+	var metadata Metadata
 
-	for _, elem := range generic {
-		// return immediately if the context is done/terminated/etc
-		select {
-		case <-ctx.Done():
-			err = multierror.Append(err, ctx.Err())
-			return nil, metadata, err
-		default:
-		}
-		// get the provider name for use in metadata
-		providerName := getProviderName(elem)
-		// now, try to open connections
+	// Return immediately if the context is done.
+	select {
+	case <-ctx.Done():
+		err = multierror.Append(err, ctx.Err())
+		return nil, metadata, err
+	default:
+	}
+
+	// Create a context with the specified timeout. This is done for backward compatibility but
+	// we should consider removing the timeout parameter alltogether given the context will
+	// container the timeout.
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	// result facilitates communication of data between the concurrent opener goroutines and
+	// the the parent goroutine.
+	type result struct {
+		ProviderName string
+		Opener       Opener
+		Err          error
+	}
+
+	// Create a channel to communicate results between opener goroutines and the parent goroutine.
+	results := make(chan result)
+
+	// Use a WaitGroup to control closing of the results channel when all opener goroutines finish.
+	var wg sync.WaitGroup
+
+	// For every provider, launch a goroutine that attempts to open a connection and report
+	// back via the results channel what happened.
+	for _, elem := range providers {
 		switch p := elem.(type) {
 		case Opener:
-			metadataLocal.ProvidersAttempted = append(metadataLocal.ProvidersAttempted, providerName)
-			tmpctx, cancel := context.WithTimeout(ctx, timeout)
-			defer cancel()
-			er := p.Open(tmpctx)
-			if er != nil {
-				err = multierror.Append(err, errors.WithMessagef(er, "provider: %v", providerName))
-				continue
-			}
-
-			opened = append(opened, elem)
-			metadataLocal.SuccessfulOpenConns = append(metadataLocal.SuccessfulOpenConns, providerName)
+			providerName := getProviderName(elem)
+			metadata.ProvidersAttempted = append(metadata.ProvidersAttempted, providerName)
+
+			wg.Add(1)
+			go func(o Opener, providerName string) {
+				defer wg.Done()
+				res := result{ProviderName: providerName, Opener: o}
+
+				if err := o.Open(ctx); err != nil {
+					res.Err = errors.WithMessagef(err, "provider: %v", providerName)
+				}
+
+				results <- res
+			}(p, providerName)
 
 		default:
-			e := fmt.Sprintf("not a Opener implementation: %T", p)
-			err = multierror.Append(err, errors.New(e))
+			err = multierror.Append(err, fmt.Errorf("not a Opener implementation: %T", p))
+		}
+	}
+
+	// Launch the goroutine to close the results channel ensuring we can exit the for-range over
+	// the results channel below.
+	go func() { wg.Wait(); close(results) }()
```

Comment:
just food for thought, if we dont use a `WaitGroup` but instead in the range over results we do a count and compare that to the `len(providers)` then we save a few allocations and use less memory. 

```bash
100	  10704031 ns/op	    2686 B/op	      46 allocs/op (with WaitGroup)
100	  10643783 ns/op	    2415 B/op	      44 allocs/op (without WaitGroup)
```

See the benchmark test i used here: https://github.com/jacobweinstock/bmclib/blob/13118c0efcf49ddbdb6e580675442fa0a79da095/bmc/connection_test.go#L63